File: app/src/main/java/knf/kuma/queue/QueueListAdapter.java
package knf.kuma.queue;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import knf.kuma.R;
import knf.kuma.database.CacheDB;
import knf.kuma.pojos.QueueObject;

public class QueueListAdapter extends RecyclerView.Adapter<QueueListAdapter.ListItemHolder> {
    private String current = "0000";
    private List<QueueObject> list = new ArrayList<>();

    @NonNull
    @Override
    public ListItemHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new ListItemHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.item_queue, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final ListItemHolder holder, int position) {
        QueueObject object = list.get(position);
        holder.chapter.setText(object.chapter.number);
        holder.icon.setImageResource(object.isFile ? R.drawable.ic_chap_down : R.drawable.ic_web);
        holder.action_delete.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                remove(holder.getAdapterPosition());
            }
        });
    }

    @Override
    public int getItemCount() {
        return list.size();
    }

    public List<QueueObject> getList() {
        return list;
    }

    public void update(String aid, List<QueueObject> list) {
        if (!current.equals(aid)) {
            current = aid;
            this.list = list;
            notifyDataSetChanged();
        }
    }

    public void remove(int position) {
        CacheDB.INSTANCE.queueDAO().remove(list.get(position));
        list.remove(position);
        notifyItemRemoved(position);
    }

    class ListItemHolder extends RecyclerView.ViewHolder {
        @BindView(R.id.chapter)
        TextView chapter;
        @BindView(R.id.icon)
        ImageView icon;
        @BindView(R.id.action_delete)
        ImageButton action_delete;

        ListItemHolder(View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/knf/kuma/queue/QueueListAdapter.kt
package knf.kuma.queue

import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import android.widget.ImageView
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import knf.kuma.R
import knf.kuma.commons.notSameContent
import knf.kuma.database.CacheDB
import knf.kuma.pojos.QueueObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.jetbrains.anko.find
import java.util.*

internal class QueueListAdapter(val callback: () -> Unit) : RecyclerView.Adapter<QueueListAdapter.ListItemHolder>() {
    private var current = "0000"
    var list: MutableList<QueueObject> = ArrayList()

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ListItemHolder {
        return ListItemHolder(LayoutInflater.from(parent.context).inflate(R.layout.item_queue, parent, false))
    }

    override fun onBindViewHolder(holder: ListItemHolder, position: Int) {
        val queueObject = list[position]
        holder.chapter.text = queueObject.chapter.number
        holder.icon.setImageResource(if (queueObject.isFile) R.drawable.ic_chap_down else R.drawable.ic_web)
        holder.actionDelete.setOnClickListener { remove(holder.adapterPosition) }
    }

    override fun getItemCount(): Int {
        return list.size
    }

    fun update(aid: String, list: MutableList<QueueObject>) {
        if (current != aid && this.list notSameContent list) {
            current = aid
            this.list = list
            notifyDataSetChanged()
        }
    }

    fun remove(position: Int) {
        if (position != -1) {
            GlobalScope.launch(Dispatchers.Main){
                withContext(Dispatchers.IO) { CacheDB.INSTANCE.queueDAO().remove(list[position]) }
                list.removeAt(position)
                notifyItemRemoved(position)
                if (list.isEmpty()) callback.invoke()
            }
        }
    }

    internal inner class ListItemHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        val chapter: TextView = itemView.find(R.id.chapter)
        val icon: ImageView = itemView.find(R.id.icon)
        val actionDelete: ImageButton = itemView.find(R.id.action_delete)
    }
}
